src/interval_tree.rs

```diff
@@ -0,0 +1,449 @@
+// Copyright (C) 2019 Alibaba Cloud. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! A special interval tree implementation for VMM resource management.
+//!
+//! It's not designed as a generic interval tree, but specialized for VMM resource management.
+//! In addition to the normal get/insert/delete/update operations, it also implements allocate/free.
+
+use std::cmp::{max, min, Ordering};
+
+/// A closed interval range [min, max].
+#[allow(missing_docs)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Hash)]
+pub struct Range {
+    pub min: u64,
+    pub max: u64,
+}
+
+impl Range {
+    /// Create a new Range object.
+    pub fn new(min: u64, max: u64) -> Self {
+        if min > max || (min == 0 && max == std::u64::MAX) {
+            panic!("interval_tree: Range({}, {}) is invalid", min, max);
+        }
+        Range { min, max }
+    }
+
+    /// Get length of the range.
+    pub fn len(&self) -> u64 {
+        self.max - self.min + 1
+    }
+
+    /// Check whether the range is empty.
+    pub fn is_empty(&self) -> bool {
+        false
+    }
+
+    /// Check whether two Range objects intersect with each other.
+    pub fn intersect(&self, other: &Range) -> bool {
+        max(self.min, other.min) <= min(self.max, other.max)
+    }
+}
+
+impl Ord for Range {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self.min.cmp(&other.min) {
+            Ordering::Equal => self.max.cmp(&other.max),
+            res => res,
+        }
+    }
+}
+
+/// Internal tree node to implement interval tree.
+#[derive(Debug)]
+struct InnerNode<T> {
+    /// Interval handled by this node.
+    key: Range,
+    /// Optional contained data, None if the node is free.
+    data: Option<T>,
```

Comment:
What is `data` designed for? I can assume the usage, for example, IO address space [start, end] is described by `key` as [start, end], ioapic irq number N is described by `key` as [N, N].

Reply:
It could be used to hold memfd for guest memory.

Reply:
Got it. Thanks for explanation.

Reply:
For other resources that don't need this `data`, what is the specific type of T?

Reply:
A simple unit struct "()" should be enough.